File: frontend/src/services/scoringService.ts
// Scoring System Service
import AsyncStorage from '@react-native-async-storage/async-storage';
import { GameSettings } from '../types';

const SCORES_KEY = '@crossword_scores';
const HIGH_SCORES_KEY = '@crossword_high_scores';

// Scoring multipliers based on difficulty
const DIFFICULTY_MULTIPLIERS = {
  easy: 1.0,
  medium: 1.5,
  hard: 2.0,
};

// Scoring multipliers based on grid size
const SIZE_MULTIPLIERS = {
  small: 1.0,    // 9x9
  medium: 1.25,  // 13x13
  large: 1.5,    // 15x15
};

// Base points
const BASE_POINTS = {
  WORD_COMPLETE: 100,        // Points for completing a word
  LETTER_CORRECT: 10,        // Points per correct letter
  NO_HINTS_BONUS: 500,       // Bonus for not using hints
  SPEED_BONUS_THRESHOLD: 300, // Seconds threshold for speed bonus
  SPEED_BONUS_MAX: 1000,     // Maximum speed bonus points
  PERFECT_SCORE_BONUS: 2000, // Bonus for completing without errors
};

export interface ScoreBreakdown {
  baseScore: number;
  difficultyBonus: number;
  sizeBonus: number;
  speedBonus: number;
  noHintsBonus: number;
  perfectBonus: number;
  totalScore: number;
  grade: 'S' | 'A' | 'B' | 'C' | 'D';
  stars: number;
}

export interface GameScore {
  id: string;
  score: number;
  breakdown: ScoreBreakdown;
  settings: GameSettings;
  timeSeconds: number;
  wordsFound: number;
  totalWords: number;
  hintsUsed: number;
  errorsCount: number;
  completedAt: number;
}

export interface HighScore {
  score: number;
  timeSeconds: number;
  completedAt: number;
}

export interface HighScores {
  easy: { small: HighScore | null; medium: HighScore | null; large: HighScore | null };
  medium: { small: HighScore | null; medium: HighScore | null; large: HighScore | null };
  hard: { small: HighScore | null; medium: HighScore | null; large: HighScore | null };
}

// Calculate score based on game performance
export function calculateScore(
  settings: GameSettings,
  timeSeconds: number,
  wordsFound: number,
  totalWords: number,
  hintsUsed: number,
  errorsCount: number
): ScoreBreakdown {
  const diffMultiplier = DIFFICULTY_MULTIPLIERS[settings.difficulty];
  const sizeMultiplier = SIZE_MULTIPLIERS[settings.size];
  
  // Base score from words found
  const baseScore = wordsFound * BASE_POINTS.WORD_COMPLETE;
  
  // Difficulty bonus
  const difficultyBonus = Math.floor(baseScore * (diffMultiplier - 1));
  
  // Size bonus
  const sizeBonus = Math.floor(baseScore * (sizeMultiplier - 1));
  
  // Speed bonus (more points for faster completion)
  let speedBonus = 0;
  if (wordsFound === totalWords && timeSeconds < BASE_POINTS.SPEED_BONUS_THRESHOLD) {
    const speedFactor = 1 - (timeSeconds / BASE_POINTS.SPEED_BONUS_THRESHOLD);
    speedBonus = Math.floor(BASE_POINTS.SPEED_BONUS_MAX * speedFactor * diffMultiplier);
  }
  
  // No hints bonus
  const noHintsBonus = hintsUsed === 0 ? Math.floor(BASE_POINTS.NO_HINTS_BONUS * diffMultiplier) : 0;
  
  // Perfect score bonus (no errors, all words found)
  const perfectBonus = (errorsCount === 0 && wordsFound === totalWords) 
    ? Math.floor(BASE_POINTS.PERFECT_SCORE_BONUS * diffMultiplier) 
    : 0;
  
  // Calculate total
  const totalScore = baseScore + difficultyBonus + sizeBonus + speedBonus + noHintsBonus + perfectBonus;
  
  // Calculate grade and stars
  const completionRate = wordsFound / totalWords;
  const { grade, stars } = calculateGrade(totalScore, completionRate, settings);
  
  return {
    baseScore,
    difficultyBonus,
    sizeBonus,
    speedBonus,
    noHintsBonus,
    perfectBonus,
    totalScore,
    grade,
    stars,
  };
}

// Calculate grade based on score and completion
function calculateGrade(
  score: number,
  completionRate: number,
  settings: GameSettings
): { grade: 'S' | 'A' | 'B' | 'C' | 'D'; stars: number } {
  const diffMultiplier = DIFFICULTY_MULTIPLIERS[settings.difficulty];
  const sizeMultiplier = SIZE_MULTIPLIERS[settings.size];
  
  // Threshold multiplier based on difficulty and size
  const thresholdMultiplier = diffMultiplier * sizeMultiplier;
  
  // Base thresholds (adjusted by multiplier)
  const thresholds = {
    S: 3000 * thresholdMultiplier,
    A: 2000 * thresholdMultiplier,
    B: 1200 * thresholdMultiplier,
    C: 600 * thresholdMultiplier,
  };
  
  let grade: 'S' | 'A' | 'B' | 'C' | 'D';
  let stars: number;
  
  if (score >= thresholds.S && completionRate === 1) {
    grade = 'S';
    stars = 5;
  } else if (score >= thresholds.A && completionRate >= 0.9) {
    grade = 'A';
    stars = 4;
  } else if (score >= thresholds.B && completionRate >= 0.7) {
    grade = 'B';
    stars = 3;
  } else if (score >= thresholds.C && completionRate >= 0.5) {
    grade = 'C';
    stars = 2;
  } else {
    grade = 'D';
    stars = 1;
  }
  
  return { grade, stars };
}

// Save a game score
export async function saveGameScore(gameScore: GameScore): Promise<void> {
  try {
    // Get existing scores
    const scoresStr = await AsyncStorage.getItem(SCORES_KEY);
    const scores: GameScore[] = scoresStr ? JSON.parse(scoresStr) : [];
    
    // Add new score
    scores.unshift(gameScore);
    
    // Keep only last 50 scores
    const trimmedScores = scores.slice(0, 50);
    
    await AsyncStorage.setItem(SCORES_KEY, JSON.stringify(trimmedScores));
    
    // Update high scores
    await updateHighScore(gameScore);
  } catch (error) {
    console.error('Error saving score:', error);
  }
}

// Update high score if applicable
async function updateHighScore(gameScore: GameScore): Promise<boolean> {
  try {
    const highScores = await getHighScores();
    const { difficulty, size } = gameScore.settings;
    
    const currentHigh = highScores[difficulty][size];
    
    if (!currentHigh || gameScore.score > currentHigh.score) {
      highScores[difficulty][size] = {
        score: gameScore.score,
        timeSeconds: gameScore.timeSeconds,
        completedAt: gameScore.completedAt,
      };
      
      await AsyncStorage.setItem(HIGH_SCORES_KEY, JSON.stringify(highScores));
      return true;
    }
    
    return false;
  } catch (error) {
    console.error('Error updating high score:', error);
    return false;
  }
}

// Get high scores
export async function getHighScores(): Promise<HighScores> {
  try {
    const highScoresStr = await AsyncStorage.getItem(HIGH_SCORES_KEY);
    if (highScoresStr) {
      return JSON.parse(highScoresStr);
    }
  } catch (error) {
    console.error('Error loading high scores:', error);
  }
  
  // Return default structure
  return {
    easy: { small: null, medium: null, large: null },
    medium: { small: null, medium: null, large: null },
    hard: { small: null, medium: null, large: null },
  };
}

// Get recent scores
export async function getRecentScores(limit: number = 10): Promise<GameScore[]> {
  try {
    const scoresStr = await AsyncStorage.getItem(SCORES_KEY);
    if (scoresStr) {
      const scores: GameScore[] = JSON.parse(scoresStr);
      return scores.slice(0, limit);
    }
  } catch (error) {
    console.error('Error loading scores:', error);
  }
  return [];
}

// Get total stats
export async function getTotalStats(): Promise<{
  totalGames: number;
  totalScore: number;
  averageScore: number;
  bestGrade: string;
}> {
  try {
    const scoresStr = await AsyncStorage.getItem(SCORES_KEY);
    if (scoresStr) {
      const scores: GameScore[] = JSON.parse(scoresStr);
      const totalGames = scores.length;
      const totalScore = scores.reduce((sum, s) => sum + s.score, 0);
      const averageScore = totalGames > 0 ? Math.floor(totalScore / totalGames) : 0;
      
      // Find best grade
      const gradeOrder = ['S', 'A', 'B', 'C', 'D'];
      let bestGrade = 'D';
      for (const score of scores) {
        if (gradeOrder.indexOf(score.breakdown.grade) < gradeOrder.indexOf(bestGrade)) {
          bestGrade = score.breakdown.grade;
        }
      }
      
      return { totalGames, totalScore, averageScore, bestGrade };
    }
  } catch (error) {
    console.error('Error loading stats:', error);
  }
  
  return { totalGames: 0, totalScore: 0, averageScore: 0, bestGrade: '-' };
}

// Format score with commas
export function formatScore(score: number): string {
  return score.toLocaleString('bg-BG');
}

// Get grade color
export function getGradeColor(grade: string): string {
  const colors: Record<string, string> = {
    'S': '#ffd700', // Gold
    'A': '#22d3ee', // Cyan
    'B': '#a855f7', // Purple
    'C': '#3b82f6', // Blue
    'D': '#6b7280', // Gray
  };
  return colors[grade] || colors['D'];
}

// Get star display
export function getStarsDisplay(stars: number): string {
  return '★'.repeat(stars) + '☆'.repeat(5 - stars);
}

File: android/app/src/main/java/com/bulgarianwordgrid/services/ScoringService.kt
package com.bulgarianwordgrid.services

import android.content.Context
import androidx.datastore.preferences.core.*
import com.bulgarianwordgrid.model.*
import kotlinx.coroutines.flow.first
import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json

object ScoringService {
    private val SCORES_KEY = stringPreferencesKey("scores")
    private val HIGH_SCORES_KEY = stringPreferencesKey("high_scores")

    private val json = Json { ignoreUnknownKeys = true }

    private val difficultyMultipliers = mapOf(
        Difficulty.EASY to 1.0,
        Difficulty.MEDIUM to 1.5,
        Difficulty.HARD to 2.0
    )

    private val sizeMultipliers = mapOf(
        GridSize.SMALL to 1.0,
        GridSize.MEDIUM to 1.25,
        GridSize.LARGE to 1.5
    )

    fun calculateScore(
        settings: GameSettings,
        timeSeconds: Int,
        wordsFound: Int,
        totalWords: Int,
        hintsUsed: Int,
        errorsCount: Int
    ): ScoreBreakdown {
        val diffMultiplier = difficultyMultipliers[settings.difficulty] ?: 1.0
        val sizeMultiplier = sizeMultipliers[settings.size] ?: 1.0

        val baseScore = wordsFound * 100
        val difficultyBonus = (baseScore * (diffMultiplier - 1)).toInt()
        val sizeBonus = (baseScore * (sizeMultiplier - 1)).toInt()

        val speedBonus = if (wordsFound == totalWords && timeSeconds < 300) {
            val speedFactor = 1.0 - (timeSeconds.toDouble() / 300.0)
            (1000 * speedFactor * diffMultiplier).toInt()
        } else 0

        val noHintsBonus = if (hintsUsed == 0) (500 * diffMultiplier).toInt() else 0
        val perfectBonus = if (errorsCount == 0 && wordsFound == totalWords) (2000 * diffMultiplier).toInt() else 0

        val totalScore = baseScore + difficultyBonus + sizeBonus + speedBonus + noHintsBonus + perfectBonus
        val completionRate = if (totalWords > 0) wordsFound.toDouble() / totalWords else 0.0
        val (grade, stars) = calculateGrade(totalScore, completionRate, settings)

        return ScoreBreakdown(
            baseScore = baseScore,
            difficultyBonus = difficultyBonus,
            sizeBonus = sizeBonus,
            speedBonus = speedBonus,
            noHintsBonus = noHintsBonus,
            perfectBonus = perfectBonus,
            totalScore = totalScore,
            grade = grade,
            stars = stars
        )
    }

    private fun calculateGrade(score: Int, completionRate: Double, settings: GameSettings): Pair<String, Int> {
        val diffMultiplier = difficultyMultipliers[settings.difficulty] ?: 1.0
        val sizeMultiplier = sizeMultipliers[settings.size] ?: 1.0
        val thresholdMultiplier = diffMultiplier * sizeMultiplier

        return when {
            score >= 3000 * thresholdMultiplier && completionRate == 1.0 -> "S" to 5
            score >= 2000 * thresholdMultiplier && completionRate >= 0.9 -> "A" to 4
            score >= 1200 * thresholdMultiplier && completionRate >= 0.7 -> "B" to 3
            score >= 600 * thresholdMultiplier && completionRate >= 0.5 -> "C" to 2
            else -> "D" to 1
        }
    }

    suspend fun saveGameScore(context: Context, gameScore: GameScore) {
        try {
            val scores = getRecentScores(context, 100).toMutableList()
            scores.add(0, gameScore)
            val trimmed = scores.take(50)
            context.gameDataStore.edit { prefs ->
                prefs[SCORES_KEY] = json.encodeToString(trimmed)
            }
            updateHighScore(context, gameScore)
        } catch (e: Exception) {
            android.util.Log.e("ScoringService", "Error saving score", e)
        }
    }

    private suspend fun updateHighScore(context: Context, gameScore: GameScore): Boolean {
        return try {
            val highScores = getHighScores(context)
            val currentHigh = highScores[gameScore.settings.size.name.lowercase()]
            if (currentHigh == null || gameScore.score > currentHigh.score) {
                val updatedMap = highScores.toMutableMap()
                updatedMap[gameScore.settings.size.name.lowercase()] = HighScore(
                    score = gameScore.score,
                    timeSeconds = gameScore.timeSeconds,
                    completedAt = gameScore.completedAt
                )
                context.gameDataStore.edit { prefs ->
                    prefs[HIGH_SCORES_KEY] = json.encodeToString(updatedMap)
                }
                true
            } else false
        } catch (e: Exception) {
            false
        }
    }

    suspend fun getHighScores(context: Context): Map<String, HighScore?> {
        return try {
            val prefs = context.gameDataStore.data.first()
            val str = prefs[HIGH_SCORES_KEY]
            if (str != null) json.decodeFromString(str) else emptyMap()
        } catch (e: Exception) {
            emptyMap()
        }
    }

    suspend fun getRecentScores(context: Context, limit: Int = 10): List<GameScore> {
        return try {
            val prefs = context.gameDataStore.data.first()
            val str = prefs[SCORES_KEY]
            if (str != null) {
                val scores: List<GameScore> = json.decodeFromString(str)
                scores.take(limit)
            } else emptyList()
        } catch (e: Exception) {
            emptyList()
        }
    }

    suspend fun getTotalStats(context: Context): Triple<Int, Int, String> {
        return try {
            val scores = getRecentScores(context, 100)
            val totalScore = scores.sumOf { it.score }
            val avgScore = if (scores.isNotEmpty()) totalScore / scores.size else 0
            val gradeOrder = listOf("S", "A", "B", "C", "D")
            val bestGrade = scores.minByOrNull { gradeOrder.indexOf(it.breakdown.grade).let { i -> if (i < 0) 4 else i } }?.breakdown.grade ?: "-"
            Triple(totalScore, avgScore, bestGrade)
        } catch (e: Exception) {
            Triple(0, 0, "-")
        }
    }

    fun formatScore(score: Int): String = "%,d".format(score)

    fun getGradeColor(grade: String): Long = when (grade) {
        "S" -> 0xFFFFD700
        "A" -> 0xFF22D3EE
        "B" -> 0xFFA855F7
        "C" -> 0xFF3B82F6
        else -> 0xFF6B7280
    }

    fun getStarsDisplay(stars: Int): String = "★".repeat(stars) + "☆".repeat(5 - stars)
}
